File: 3695.maximize-alternating-sum-using-swaps.py
#
# @lc app=leetcode id=3695 lang=python3
#
# [3695] Maximize Alternating Sum Using Swaps
#

# @lc code=start
class Solution:
    def maxAlternatingSum(self, nums: List[int], swaps: List[List[int]]) -> int:
        # Initialize union-find structures
        parent = list(range(len(nums)))
        rank = [0] * len(nums)

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            rootX = find(x)
            rootY = find(y)
            if rootX != rootY:
                if rank[rootX] > rank[rootY]:
                    parent[rootY] = rootX
                elif rank[rootX] < rank[rootY]:
                    parent[rootX] = rootY
                else:
                    parent[rootY] = rootX
                    rank[rootX] += 1
        
        # Connect nodes according to swaps
        for pi, qi in swaps:
            union(pi, qi)
        
        # Group elements by their connected component roots
        from collections import defaultdict as ddict
        components = ddict(list)
        for i in range(len(nums)):
            components[find(i)].append(i)
        
        # Calculate maximum alternating sum by optimizing each component separately
        max_alternating_sum = 0
        for comp in components.values():
            # Sort values within this component to maximize alternating sum contribution
            numbers = sorted((nums[i] for i in comp), reverse=True)
            comp_sum = 0 
            # Calculate alternating sum for this sorted order 
            for idx, num in enumerate(numbers): 
                if idx % 2 == 0: 
                    comp_sum += num 
                else: 
                    comp_sum -= num 
            max_alternating_sum += comp_sum 
       
       # Return total maximum alternating sum across all components 
       return max_alternating_sum \\ @lc code=end

File: 3695.maximize-alternating-sum-using-swaps.java
#
# @lc app=leetcode id=3695 lang=java
#
# [3695] Maximize Alternating Sum Using Swaps
#
# @lc code=start
import java.util.*;
class Solution {
    public long maxAlternatingSum(int[] nums, int[][] swaps) {
        int n = nums.length;
        DSU dsu = new DSU(n);
        for (int[] swap : swaps) {
            dsu.union(swap[0], swap[1]);
        }
        Map<Integer, List<Integer>> components = new HashMap<>();
        for (int i = 0; i < n; i++) {
            int root = dsu.find(i);
            components.computeIfAbsent(root, k -> new ArrayList<>()).add(i);
        }
        long res = 0;
        for (List<Integer> comp : components.values()) {
            List<Integer> vals = new ArrayList<>();
            for (int i : comp) vals.add(nums[i]);
            Collections.sort(vals, Collections.reverseOrder());
            Collections.sort(comp);
            for (int j = 0; j < comp.size(); j++) {
                if (comp.get(j) % 2 == 0) res += vals.get(j);
                else res -= vals.get(j);
            }
        }
        return res;
    }
}
class DSU {
    int[] parent;
    public DSU(int n) {
        parent = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;
    }
    public int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
    public void union(int x, int y) {
        parent[find(x)] = find(y);
    }
}
# @lc code=end